test/lit/passes/dae-gc.wast

```diff
@@ -293,4 +294,224 @@
+  (local $temp anyref)
+  (local.set $temp (call $refine-return-flow))
+
+  (ref.null func)
+ )
+ ;; CHECK:      (func $call-refine-return-flow (result funcref)
+ ;; CHECK-NEXT:  (local $temp anyref)
+ ;; CHECK-NEXT:  (local.set $temp
+ ;; CHECK-NEXT:   (call $call-refine-return-flow)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (if (result funcref)
+ ;; CHECK-NEXT:   (i32.const 1)
+ ;; CHECK-NEXT:   (call $refine-return-flow)
+ ;; CHECK-NEXT:   (call $refine-return-flow)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $call-refine-return-flow (result anyref)
+  (local $temp anyref)
+  (local.set $temp (call $call-refine-return-flow))
+
+  ;; After refining the return value of the above function, refinalize will
+  ;; update types here, which will lead to updating the if, and then the entire
+  ;; function's return value.
+  (if (result anyref)
+   (i32.const 1)
+   (call $refine-return-flow)
+   (call $refine-return-flow)
+  )
+ )
+
+ ;; Refine the return type based on a return.
+ ;; CHECK:      (func $refine-return-return (result funcref)
+ ;; CHECK-NEXT:  (local $temp anyref)
+ ;; CHECK-NEXT:  (local.set $temp
+ ;; CHECK-NEXT:   (call $refine-return-return)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (return
+ ;; CHECK-NEXT:   (ref.null func)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $refine-return-return (result anyref)
+  (local $temp anyref)
+  (local.set $temp (call $refine-return-return))
+
+  (return (ref.null func))
+ )
+
+ ;; Refine the return type based on multiple values.
+ ;; CHECK:      (func $refine-return-many (result funcref)
+ ;; CHECK-NEXT:  (local $temp anyref)
+ ;; CHECK-NEXT:  (local.set $temp
+ ;; CHECK-NEXT:   (call $refine-return-many)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (if
+ ;; CHECK-NEXT:   (i32.const 1)
+ ;; CHECK-NEXT:   (return
+ ;; CHECK-NEXT:    (ref.null func)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (if
+ ;; CHECK-NEXT:   (i32.const 2)
+ ;; CHECK-NEXT:   (return
+ ;; CHECK-NEXT:    (ref.null func)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (ref.null func)
+ ;; CHECK-NEXT: )
+ (func $refine-return-many (result anyref)
+  (local $temp anyref)
+  (local.set $temp (call $refine-return-many))
+
+  (if
+   (i32.const 1)
+   (return (ref.null func))
+  )
+  (if
+   (i32.const 2)
+   (return (ref.null func))
+  )
+  (ref.null func)
+ )
+
+ ;; CHECK:      (func $refine-return-many-partial (result anyref)
+ ;; CHECK-NEXT:  (local $temp anyref)
+ ;; CHECK-NEXT:  (local.set $temp
+ ;; CHECK-NEXT:   (call $refine-return-many-partial)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (if
+ ;; CHECK-NEXT:   (i32.const 1)
+ ;; CHECK-NEXT:   (return
+ ;; CHECK-NEXT:    (ref.null func)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (if
+ ;; CHECK-NEXT:   (i32.const 2)
+ ;; CHECK-NEXT:   (return
+ ;; CHECK-NEXT:    (ref.null data)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (ref.null func)
+ ;; CHECK-NEXT: )
+ (func $refine-return-many-partial (result anyref)
```

Comment:
"partial" makes it sound like some refinement will still happen, but it doesn't in these tests. Maybe "blocked" or similar instead?

Reply:
Done.